File: Zigzag Conversion/ZigzagConversion.py
def convert(s: str, numRows: int) -> str:
    if numRows == 1:
        return s

    ans = [''] * numRows
    crow = 0
    down = False

    for c in s:
        ans[crow] += c

        if crow == 0:
            down = True
        elif crow == numRows - 1:
            down = False

        crow += 1 if down else -1

    return "".join(ans)

# Test
s = "PAYPALISHIRING"
numRows = 3
print(convert(s, numRows))  # Output: "PAHNAPLSIIGYIR"

File: Zigzag Conversion/ZigzagConversion.java
import java.util.*;

public class ZigzagConversion {
    public String convert(String s, int numRows) {
        if (numRows == 1) {
            return s;
        }

        List<StringBuilder> rows = new ArrayList<>();
        for (int i = 0; i < numRows; i++) {
            rows.add(new StringBuilder());
        }

        int crow = 0;
        boolean down = false;

        for (char c : s.toCharArray()) {
            rows.get(crow).append(c);
            if (crow == 0) {
                down = true;
            } else if (crow == numRows - 1) {
                down = false;
            }

            crow += down ? 1 : -1;
        }

        StringBuilder result = new StringBuilder();
        for (StringBuilder row : rows) {
            result.append(row);
        }

        return result.toString();
    }

    public static void main(String[] args) {
        ZigzagConversion zigzag = new ZigzagConversion();
        String result = zigzag.convert("PAYPALISHIRING", 3);
        System.out.println(result); 
    }
}
